src/Key.ts

```diff
@@ -0,0 +1,172 @@
+]);
+
+/**
+ * Code comes in many formats.
+ * Keys start with Key (ex: KeyN)
+ * Digits start with Digit (ex: Digit1)
+ * Other characters like .,[] are all treated with special string values.
+ */
+function getCodeValue(code: string): string | undefined {
+  if (code.startsWith('Key')) {
+    return code.substring(3, code.length).toLowerCase();
+  }
+
+  if (code.startsWith('Digit')) {
+    return code.substring(5, code.length).toLowerCase();
+  }
+
+  return codeMap.get(code);
+}
+
+/**
+ * Gets the key code value.
+ */
+function getKeyCodeValue(keyCode: number): string | undefined {
+  if (keyCodeMap.has(keyCode)) {
+    return keyCodeMap.get(keyCode);
+  }
+
+  return String.fromCharCode(keyCode).toLowerCase();
+}
+
+/**
+ * https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode
+ * Value detection is hard. Basically we want to use code if we can but code comes in many formats.
+ * Also code is not garuntueed in all browsers so we need to fallback to keyCode or key.
+ */
+function getKeyValue(event: KeyboardEvent): string {
+  if (event.code) {
+    const value = getCodeValue(event.code);
+    if (value) {
+      return value;
+    }
+  }
+
+  if (event.keyCode) {
+    const keyCodeValue = getKeyCodeValue(event.keyCode);
+    if (keyCodeValue) {
+      return keyCodeValue;
+    }
+  }
+
+  return event.key;
+}
+
+/**
+ * Key class for converting and handling key events.
+ */
+class Key {
+  /** The key (character) that was pressed */
+  public key: string;
+  /** If the alt key was pressed */
+  public alt?: boolean;
+  /** If the ctrl key was pressed */
+  public ctrl?: boolean;
+  /** If the shift key was pressed */
+  public shift?: boolean;
+  /** If the meta key was pressed */
+  public meta?: boolean;
+
+  /** Converts a browser KeyboardEvent into a Key */
+  public static fromKeyEvent(event: KeyboardEvent): Key {
+    const options = {
+      alt: event.altKey,
+      ctrl: event.ctrlKey,
+      shift: event.shiftKey,
+      meta: event.metaKey
+    };
+
+    return new Key(getKeyValue(event), options);
+  }
+
+  constructor(key: string, options?: KeyModifierOptions) {
+    this.key = key.toLowerCase();
+    this.alt = (options && options.alt) || false;
+    this.ctrl = (options && options.ctrl) || false;
+    this.shift = (options && options.shift) || false;
+    this.meta = (options && options.meta) || false;
+  }
+
+  /** Creates a string representation of the key */
+  public serialize(): string {
+    let retVal = this.key || '';
+
+    if (this.ctrl) {
+      retVal += 'ctrl';
+    }
+
+    if (this.alt) {
+      retVal += 'alt';
+    }
+
+    if (this.shift) {
+      retVal += 'shift';
+    }
+
+    if (this.meta) {
+      retVal += 'meta';
+    }
+
+    return retVal;
+  }
+
+  /** Tests equalality of another key. */
+  public equals(key: Key): boolean {
+    return key.serialize() === this.serialize();
+  }
+}
+
+/** Data structure representing a native key event. */
+interface NativeKey {
```

Comment:
Do we really need to send `charCode`, `code` and `keyCode`? What are they used for?

Reply:
I ask because if we don't actually need to send those, the `Key` class is largely unnecessary. At that point it's just renaming fields and handling a comparison, which could easily just be a helper function.

Reply:
I think I mentioned this in a different section of this review. I will copy it over here. Let me know if you have any additional feedback on this.

> I could be missing something here but I think code and keyCode are pretty important. I agree that most of the cases you could blame alt as the mapping key but I think that is a valid use case.
> 
> The following are some examples of confusing behavior with just those properties `alt+f`, `alt+n` and `alt+e`. The big issue is with `alt+n` and `alt+e`. In the case of windows you would get key = n/e alt = true. In the case of mac you would get key = dead alt = true for both. The only real way to distinguish these two events on mac is using code (KeyN/KeyE) or keyCode (732/180)
> 
> The way I have it now seems like it should handle all cases pretty well. You register the key using its raw value (e/f/n) and any modifiers; However, when you get the event back from the API it provides key in its raw form along with code and keyCode to help you decide how to handle that event.

Reply:
Misread this. I see so you would like to get rid of the class. Yeah I could look into that.

Reply:
Key class is now gone in favor of a helper function. Let me know if there are any other requested changes or feedback